common/lib/host_list_provider/monitoring/monitoring_host_list_provider.ts

```diff
@@ -0,0 +1,123 @@
+import { ClusterTopologyMonitor, ClusterTopologyMonitorImpl } from "./cluster_topology_monitor";
+import { PluginService } from "../../plugin_service";
+import { HostListProviderService } from "../../host_list_provider_service";
+import { ClientWrapper } from "../../client_wrapper";
+import { DatabaseDialect } from "../../database_dialect/database_dialect";
+import { AwsWrapperError } from "../../utils/errors";
+import { Messages } from "../../utils/messages";
+import { WrapperProperties } from "../../wrapper_property";
+
+export class MonitoringRdsHostListProvider extends RdsHostListProvider {
+  static readonly CACHE_CLEANUP_NANOS: bigint = BigInt(60_000_000_000); // 1 minute.
+  static readonly MONITOR_EXPIRATION_NANOS: bigint = BigInt(15 * 60_000_000_000); // 15 minutes.
+  static readonly DEFAULT_TOPOLOGY_QUERY_TIMEOUT_MS = 5000; // 5 seconds.
+
+  private static monitors: SlidingExpirationCache<string, ClusterTopologyMonitorImpl> = new SlidingExpirationCache(
+    MonitoringRdsHostListProvider.CACHE_CLEANUP_NANOS,
+    () => true,
+    async (monitor: ClusterTopologyMonitorImpl) => {
+      try {
+        await monitor.close();
+      } catch {
+        // Ignore.
+      }
+    }
+  );
+
+  private readonly pluginService: PluginService;
+
+  constructor(properties: Map<string, any>, originalUrl: string, hostListProviderService: HostListProviderService, pluginService: PluginService) {
+    super(properties, originalUrl, hostListProviderService);
+    this.pluginService = pluginService;
+  }
+
+  async clearAll(): Promise<void> {
+    RdsHostListProvider.clearAll();
+    for (const [key, monitor] of MonitoringRdsHostListProvider.monitors.entries) {
+      await monitor.item.close();
+    }
+    MonitoringRdsHostListProvider.monitors.clear();
+  }
+
+  async queryForTopology(targetClient: ClientWrapper, dialect: DatabaseDialect): Promise<HostInfo[]> {
+    let monitor: ClusterTopologyMonitor = MonitoringRdsHostListProvider.monitors.get(
+      this.clusterId,
+      MonitoringRdsHostListProvider.MONITOR_EXPIRATION_NANOS
+    );
+    if (!monitor) {
+      monitor = this.initMonitor();
+    }
+
+    try {
+      return await monitor.forceRefresh(targetClient, MonitoringRdsHostListProvider.DEFAULT_TOPOLOGY_QUERY_TIMEOUT_MS);
+    } catch {
+      return null;
+    }
+  }
+
+  async sqlQueryForTopology(targetClient: ClientWrapper): Promise<HostInfo[]> {
+    const dialect: DatabaseDialect = this.hostListProviderService.getDialect();
+    if (!this.isTopologyAwareDatabaseDialect(dialect)) {
+      throw new TypeError(Messages.get("RdsHostListProvider.incorrectDialect"));
+    }
+    return await dialect.queryForTopology(targetClient, this).then((res: any) => this.processQueryResults(res));
+  }
+
+  async getWriterId(targetClient: ClientWrapper): Promise<string> {
+    const hostInfo: HostInfo = await this.identifyConnection(targetClient, this.hostListProviderService.getDialect());
```

Comment:
`identifyConnection` gets the host name of the current connection, which might not be a writer instance, so I don't think this will reliably return the writer ID.

Reply:
Check the `IS_WRITER_QUERY` variable in JDBC